src/audio/smart_amp/smart_amp.c

```diff
@@ -0,0 +1,662 @@
+
+	buffer_unlock(sad->source_buf, source_flags);
+	buffer_unlock(sad->sink_buf, sink_flags);
+
+	avail_frames = avail_passthrough_frames;
+
+	buffer_lock(sad->feedback_buf, &feedback_flags);
+	if (sad->feedback_buf->source->state == dev->state) {
+		/* feedback */
+		avail_feedback_frames = sad->feedback_buf->stream.avail /
+			audio_stream_frame_bytes(&sad->feedback_buf->stream);
+
+		avail_frames = MIN(avail_passthrough_frames,
+				   avail_feedback_frames);
+
+		feedback_bytes = avail_frames *
+			audio_stream_frame_bytes(&sad->feedback_buf->stream);
+
+		buffer_unlock(sad->feedback_buf, feedback_flags);
+
+		comp_dbg(dev, "smart_amp_copy(): processing %d feedback frames (avail_passthrough_frames: %d)",
+			 avail_frames, avail_passthrough_frames);
+
+		sad->process(dev, &sad->feedback_buf->stream,
+			     &sad->sink_buf->stream, avail_frames,
+			     sad->config.feedback_ch_map, true);
+
+		comp_update_buffer_consume(sad->feedback_buf, feedback_bytes);
+	}
+
+	/* bytes calculation */
+	buffer_lock(sad->source_buf, &source_flags);
+	source_bytes = avail_frames *
+		audio_stream_frame_bytes(&sad->source_buf->stream);
+	buffer_unlock(sad->source_buf, source_flags);
+
+	buffer_lock(sad->sink_buf, &sink_flags);
+	sink_bytes = avail_frames *
+		audio_stream_frame_bytes(&sad->sink_buf->stream);
+	buffer_unlock(sad->sink_buf, sink_flags);
+
+	/* process data */
+	sad->process(dev, &sad->source_buf->stream, &sad->sink_buf->stream,
+		     avail_frames, sad->config.source_ch_map, false);
+
+	/* source/sink buffer pointers update */
+	comp_update_buffer_consume(sad->source_buf, source_bytes);
+	comp_update_buffer_produce(sad->sink_buf, sink_bytes);
+
+	return ret;
+}
+
+static int smart_amp_reset(struct comp_dev *dev)
+{
+	comp_dbg(dev, "smart_amp_reset()");
+
+	comp_set_state(dev, COMP_TRIGGER_RESET);
+
+	return 0;
+}
+
+static int smart_amp_prepare(struct comp_dev *dev)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct comp_buffer *source_buffer;
+	struct list_item *blist;
+	int ret;
+
+	comp_dbg(dev, "smart_amp_prepare()");
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	/* searching for stream and feedback source buffers */
+	list_for_item(blist, &dev->bsource_list) {
+		source_buffer = container_of(blist, struct comp_buffer,
+					     sink_list);
+
+		if (source_buffer->source->comp.type == SOF_COMP_DEMUX)
+			sad->feedback_buf = source_buffer;
+		else
+			sad->source_buf = source_buffer;
+	}
+
+	sad->sink_buf = list_first_item(&dev->bsink_list, struct comp_buffer,
+					source_list);
+
+	sad->in_channels = sad->source_buf->stream.channels;
+	sad->out_channels = sad->sink_buf->stream.channels;
+
+	sad->feedback_buf->stream.channels = sad->config.feedback_channels;
+
+	/* TODO:
+	 * ATM feedback buffer frame_fmt is hardcoded to s32_le. It should be
+	 * removed when parameters negotiation between pipelines will prepared
+	 */
+	sad->feedback_buf->stream.frame_fmt = SOF_IPC_FRAME_S32_LE;
```

Comment:
should we return error on streams that are not S32LE so we don't have silent memory failures?

Reply:
There were two reasons why S32_LE format is hardcoded here:
1. At the beginning there was requirement only for that format on feedback stream.
2. Our pipeline params propagation mechanism does not handle case with DSM topology (format parameter is not properly propagated to feedback buffer, so it was hardcoded at that moment (assuming that we only need s32_le)).

If we want to delete this hardcode, we should reopen @keyonjie PR with params negotiation #2788.

Reply:
If the current platforms only support S32LE then i am fine with the hardcode but I recommend we open a tracking bug in case other formats such float wish to use this then we have a bug we can refer to

Reply:
@bkokoszx Could you please help on this?

Reply:
@ryans-lee can you put a check in here that returns error with some trace error output so this does not silenttly fail. You should keep the TODO info so that this can be addressed one the negotion is completed by @keyonjie .

Reply:
@lgirdwood I've added a check and put 'com_err' message to notify the issue. I considered to return error in smart_amp_process, but it causes several errors and blocked testing for capture stream.